### ISSUE TITLE
Deprecate `--parallel` in favor of the configuration cache

### ISSUE THREAD

**Issue by adammurdoch** [gradle/gradle#25317]
### Expected Behavior

The `--parallel` option is now superseded by the configuration cache, which is able to run all tasks in parallel and offers additional benefits. We should retire `--parallel` and help people use the configuration cache instead.

### Current Behavior (optional)

There are currently 2 separate features that attempt to solve the same problem. This leads to unnecessary extra work for everyone, such as the cost of maintaining, documenting and testing these features or the cost of a build engineer understanding the difference between the two options and the shortcomings of `--parallel`.

### Context

n/a

**Comment by snazy** [comment]
Just noticed that `--parallel` is about to be deprecated. This is a bit sad, because a couple of plugins are not yet ready to use the configuration cache - speaking for the Quarkus Gradle plugin. There are some things that don't work well with the configuration cache, so all tasks from the Quarkus 3 Gradle plugin have to disable the configuration cache.

Wonder whether `--parallel` is "marked for removal" or "just deprecated" but will stay around.

**Comment by snazy** [comment]
Update: Gradle 8.2 works with the Quarkus 3 Gradle plugin, but not w/ the Quarkus 2 Gradle plugin.

**Comment by 3flex** [comment]
This might be a reason to include the deprecation in 8.x and remove in 9: https://github.com/gradle/gradle/issues/10385#issuecomment-1684664558

**Comment by matejdro** [comment]
So, when configuration cache is enabled, all tasks will automatically run in parallel, without `--parallel`?

**Comment by Vampire** [comment]
Yes, see: https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:intro:performance_improvements

**Comment by alexei-osipov** [comment]
There is wast amount of plugins that do no support "Configuration Cache" mechanism. Also, "Configuration Cache" is much more restrictive mode of Gradle operation. So "--parallel" mode is a good option for projects that can't switch to "Configuration Cache" yet. Removal of parallel mode in Gradle 9 may mean that many users will be stuck on Gradle 8.

**Comment by Vampire** [comment]
What makes you think it will be removed in Gradle 9?
The deprecation of it (this ticket) does not even have a milestone set. 

Besides that, according to https://github.com/gradle/build-tool-roadmap/issues/75 the current plan is to make configuration cache the default in Gradle 9 and deprecate running without configuration cache in Gradle 10, so latest then (resp. for 11) builds have to adopt it anyway to upgrade. 🤷‍♂️

**Comment by tkrah** [comment]
> Yes, see: https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:intro:performance_improvements

But the docs there has this:

> Parallel Task Execution: **When** parallel execution is enabled

So if I do not enable parallel mode, it should not run in parallel - where it is written, that if you use the CC, that this implies **--parallel** ?

**Comment by Vampire** [comment]
I think the "When" is a documentation bug.
Afair, you cannot disable parallel execution in general when using configuration cache.
If you for example have
```kotlin
val foo by tasks.registering {
    doLast {
        println("${LocalDateTime.now()} - Foo 1")
        Thread.sleep(3000)
        println("${LocalDateTime.now()} - Foo 2")
    }
}
val bar by tasks.registering {
    doLast {
        println("${LocalDateTime.now()} - Bar 1")
        Thread.sleep(3000)
        println("${LocalDateTime.now()} - Bar 2")
    }
}
```
then even with `gradle foo bar --configuration-cache --no-parallel` those tasks are running in parallel, even if their output is grouped per task (hence the date-time output.

It makes sense to use `--parallel` together with `--configuration-cache` in case the CC cannot be used due to CC problems,
so that at least tasks in different projects can run in parallel even if CC is not used.
But `--no-parallel` should not have any effect if CC is actually used.

If you want tasks to not run in parallel, you need a shared build service - even one without any logic - and then configure the tasks that should not run in parallel to use that service and configure the service to be usable by at most one task in parallel. This will ensure those tasks will not run in parallel.

**Comment by tkrah** [comment]
> I think the "When" is a documentation bug. Afair, you cannot disable parallel execution in general when using configuration cache. If you for example have
> 
> val foo by tasks.registering {
>     doLast {
>         println("${LocalDateTime.now()} - Foo 1")
>         Thread.sleep(3000)
>         println("${LocalDateTime.now()} - Foo 2")
>     }
> }
> val bar by tasks.registering {
>     doLast {
>         println("${LocalDateTime.now()} - Bar 1")
>         Thread.sleep(3000)
>         println("${LocalDateTime.now()} - Bar 2")
>     }
> }
> 
> then even with `gradle foo bar --configuration-cache --no-parallel` those tasks are running in parallel, even if their output is grouped per task (hence the date-time output.
> 
> It makes sense to use `--parallel` together with `--configuration-cache` in case the CC cannot be used due to CC problems, so that at least tasks in different projects can run in parallel even if CC is not used. But `--no-parallel` should not have any effect if CC is actually used.
> 
> If you want tasks to not run in parallel, you need a shared build service - even one without any logic - and then configure the tasks that should not run in parallel to use that service and configure the service to be usable by at most one task in parallel. This will ensure those tasks will not run in parallel.

I have enabled CC and my tasks do not run in parallel - how did I do that?

And imho that is a critical documentation bug if that **when** is wrong and its runs parallel by default, because I would have been forced to use shared build services which I just have on my bucket list to write, although CC is enabled (and beneficial already, even if it is not running in parallel at the moment).

Hm maybe because I also have:

org.gradle.workers.max=1

set - but the docs https://docs.gradle.org/current/userguide/build_environment.html#gradle_properties_reference says, this will only have a effect if parallel is true - which you said does have no effect here at all (https://github.com/gradle/gradle/issues/37763) - so why should that matter - or does it matter per coincidence here (Seems I am going to check the sources, I am interested now ^^).

**Comment by tkrah** [comment]
> 
> I have enabled CC and my tasks do not run in parallel - how did I do that?
> 
> And imho that is a critical documentation bug if that **when** is wrong and its runs parallel by default, because I would have been forced to use shared build services which I just have on my bucket list to write, although CC is enabled (and beneficial already, even if it is not running in parallel at the moment).
> 
> Hm maybe because I also have:
> 
> org.gradle.workers.max=1
> 
> set - but the docs https://docs.gradle.org/current/userguide/build_environment.html#gradle_properties_reference says, this will only have a effect if parallel is true - which you said does have no effect here at all ([#37763](https://github.com/gradle/gradle/issues/37763)) - so why should that matter - or does it matter per coincidence here (Seems I am going to check the sources, I am interested now ^^).

To answer myself - yeah it is:

`org.gradle.workers.max=1`

which does disable the parallel runs - although `org.gradle.parallel=false` is set, having `org.gradle.workers.max=4` will run stuff in parallel. So the docs here are really misleading BUT to the rescue, setting those workers to **1** will help to have it serialized nevertheless.

**Comment by Vampire** [comment]
Ah, yeah, if you have the hand break on, this will indeed also cause serial execution.

But I would really use a build service, to at least let other workers run in parallel.
All you need to serialize all tasks is
```kotlin
interface TaskSemaphore : BuildService<None>
val taskSemaphore = project.gradle.sharedServices.registerIfAbsent("taskSemaphore", TaskSemaphore::class) {
    maxParallelUsages = 1
}
tasks.configureEach {
    usesService(taskSemaphore)
}
```

**Comment by tkrah** [comment]
> Ah, yeah, if you have the hand break on, this will indeed also cause serial execution.
> 
> But I would really use a build service, to at least let other workers run in parallel. All you need to serialize all tasks is
> 
> interface TaskSemaphore : BuildService<None>
> val taskSemaphore = project.gradle.sharedServices.registerIfAbsent("taskSemaphore", TaskSemaphore::class) {
>     maxParallelUsages = 1
> }
> tasks.configureEach {
>     usesService(taskSemaphore)
> }

Thanks for the sample, like I said, it is on my bucket list.
My first naive thought to use it like that in the past did not work, still need to figure out how task finalizers and prerequisites (mustRunBefore, mustRunAfter) must be handled in that case, I will figure it out though (and maybe I will ask in the Gradle forums or Slack community about that too ;-) ).

**Comment by Vampire** [comment]
No need to handle them, those only control whether and in which order things are executed.
The snippet I gave just makes sure no task runs in parallel.
